### DISCUSSION TITLE
Special token support

### DISCUSSION THREAD

**Discussion by Dampfinchen** [ggml-org/llama.cpp#3466] [Ideas]
A few days ago, Open Orca released a new model called Mistral-7B-Openorca. This uses the ChatML format which has <|im_end|> as a special EOS token that is currently not recognized by llama.cpp

<img width="477" alt="image" src="https://github.com/ggerganov/llama.cpp/assets/59751859/a205a79b-0542-4c8b-bdb7-06f5ad9bf55a">

This is the format in question.

https://github.com/openai/openai-python/blob/main/chatml.md

model 

https://huggingface.co/Open-Orca/Mistral-7B-OpenOrca/commit/17572416df27482d71dda9ea6bdea1733d8cee5d

**Comment by TechnotechGit** [comment]
This might be better suggested as an issue? If you use llama cpp python, I think you should be able to get over the issue by using the actual hf tokeniser using something like [huggingface/tokenizers](https://github.com/huggingface/tokenizers) or simply [transformers](https://github.com/huggingface/transformers) itself, but it would be nice to have a solution to this baked into llama.cpp imo.

**Comment by staviq** [comment]
This has already been noticed, and couple other problems with mistral-openorca too, see #3454, #3455 and #346

Quick and dirty solution and "proper" prompt format parameters: https://github.com/ggerganov/llama.cpp/pull/3455#issuecomment-1745720746
